subsys/nrf_compress/src/lzma.c

```diff
@@ -104,15 +137,89 @@ static CLzma2Dec lzma_decoder;
 static CLzmaDec lzma_decoder;
 #endif
 
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+/**
+ * @brief Synchronize dictionary cache with external dictionary.
+ *
+ * This function synchronizes data in cache with external
+ * dictionary and proceeds with cache window.
+ *
+ * @param handle pointer to Lzma dictionary handle struct, for dictionary size reference.
+ *
+ * @retval 0 on successful synchronization
+ * @retval -EIO on any error with reading/writing to external dictionary
+ *
+ */
+static int synchronize_cache(const DictHandle *handle)
+{
+	SizeT const dict_write_size = cache.dict_pos_end - cache.dict_pos_begin + 1;
+
+	if (ext_dict->write(cache.dict_pos_begin,
+			cache.data, dict_write_size) != dict_write_size) {
+		return -EIO;
+	}
+
+	cache.dict_pos_begin = cache.dict_pos_end + 1;
+
+	if (cache.dict_pos_begin == handle->dicBufSize) {
+		/** We reached the end of dictionary, start caching from the beginning.*/
+		cache.dict_pos_begin = 0;
+	}
+
+	SizeT const dict_read_size =
+			(handle->dicBufSize - cache.dict_pos_begin) < sizeof(cache.data) ?
+				(handle->dicBufSize - cache.dict_pos_begin)
+				: sizeof(cache.data);
+
+	cache.dict_pos_end = cache.dict_pos_begin + dict_read_size - 1;
+
+	if (ext_dict->read(cache.dict_pos_begin,
+			cache.data, dict_read_size) != dict_read_size) {
+		return -EIO;
+	}
+
+	cache.invalid = false;
+
+	return 0;
+}
+#endif
+
+/**
+ * @brief Check the instance of lzma_codec during API calls.
+ */
+static int check_inst(void *inst)
+{
+#if defined(CONFIG_NRF_COMPRESS_EXTERNAL_DICTIONARY)
+	if (inst != CONTAINER_OF(ext_dict, lzma_codec, dict_if)) {
+		return -EINVAL;
+	}
+#else
+	ARG_UNUSED(inst);
+#endif
+	return 0;
+}
+
 static int lzma_reset(void *inst);
 
 static int lzma_init(void *inst)
 {
 	int rc = 0;
 
+#if defined(CONFIG_NRF_COMPRESS_EXTERNAL_DICTIONARY)
+	if (inst == NULL)
+		return -EINVAL;
```

Comment:
brackets for all if checks, @carlescufi I thought a check for this was added which would fail CI?